### PR TITLE
chore(core): pre provision check for vs

### PR DESCRIPTION
**Pull request by jayzhang**

### CODE REVIEW

File: packages/fx-core/src/component/coordinator/index.ts [hunk 1]
```diff
@@ -290,6 +291,69 @@ export class Coordinator {
...
+    const projectModel = maybeProjectModel.value;
+    const cycles: ILifecycle[] = [
+      projectModel.registerApp,
+      projectModel.provision,
+      projectModel.configureApp,
+    ].filter((c) => c !== undefined) as ILifecycle[];
+
+    // 2. check each cycle
+    for (const cycle of cycles) {
+      cycle.resolvePlaceholders();
```
Comment (1yefuwang1) [inline packages/fx-core/src/component/coordinator/index.ts:336]: Maybe it's better to check the return value of `resolvePlaceholders()` to get unresolved placeholders,

Comment (jayzhang) [reply packages/fx-core/src/component/coordinator/index.ts:336]: There is no need to check whether placeholders are resolved or not, we only need the final value

File: packages/fx-core/src/component/coordinator/index.ts [hunk 2]
```diff
@@ -290,6 +291,69 @@ export class Coordinator {
...
+      for (const driver of cycle.driverDefs) {
+        if (AzureActions.includes(driver.uses)) {
+          res.needAzureLogin = true;
+        }
+        if (M365Actions.includes(driver.uses)) {
+          res.needM365Login = true;
+        }
+      }
+      const firstDriver = cycle.driverDefs[0] || undefined;
```
Comment (1yefuwang1) [inline packages/fx-core/src/component/coordinator/index.ts:345]: It's possible that driverDefs is an empty array, in which case `cycle.driverDefs[0]` yields `undefined`

**Review by xzf0587 — DISMISSED** [review]

File: packages/fx-core/src/component/coordinator/index.ts (hunk 2 above)
Comment (jayzhang) [reply packages/fx-core/src/component/coordinator/index.ts:345]: cycle.driverDefs[0] === undefined is ok for the following code to run correctly

Comment (jayzhang) [reply packages/fx-core/src/component/coordinator/index.ts:345]: image

**Review by LongOddCode — APPROVED** [review]

**Review by 1yefuwang1 — APPROVED** [review]

**Pull request merged**